site/content/en/docs/Guides/local-game-server.md

```diff
@@ -9,6 +9,8 @@ description: >
 
 You can register a local game server with Agones. This means you can run an experimental build of your game server in the Agones environment without the need of packaging and deploying it to a fleet. This allows you to quickly iterate on your game server code while still being able to plugin to your Agones environment.
 
+This can be used in combination with a [local SDK Server]({{< ref "./Client SDKs/local.md" >}}) in ["out of cluster" mode]({{< ref "./Client SDKs/local.md#running-locally-using-out-of-cluster-mode" >}}).
```

Comment:
So when reading the "out of cluster mode" - I automatically assumed I could connect to my local game server binary.... but reading this, it doesn't seem to be the case.

I'm wondering if there's a way we can combine it into one whole process? 🤔 not entirely sure how, but it feels like it's worth exploring. WDYT?

Reply:
>  I automatically assumed I could connect to my local game server binary

That is the case. This is what I do for local testing. I have my custom game server running in Visual Studio, I run the SDK Server on a terminal, and I have it connect to my cluster. This allows me to allocate the game server via normal prod flows (e.g. external calls onto the allocator service) which are then fully integrated with my breakpoint-able game server in VS.

> I'm wondering if there's a way we can combine it into one whole process?

I mean, the steps to do so are mentioned in documentation. One would need to call `kubectl apply -f dev-gameserver.yaml`, then run `./sdk-server.linux.amd64 ...`, then run their custom game server binary. Everything is then connected.

Reply:
A few thoughts on this:

1. From reading the documentation, the combination of the two approaches wasn't immediately obvious to me. I'm _almost_ wondering if this would be better as a part of https://agones.dev/site/docs/guides/local-game-server/ than here 🤔  (at the very least, they should probably link to each other though).
2. We should be super clear that you need both to actually connect. With the current docs it's easy to assume you can and you may not realise until the very end that you can't (or even that the `agones.dev/dev-address` GameSever approach exists -- which is kinda why I'm thinking we may want this in https://agones.dev/site/docs/guides/local-game-server/ as well, since one really does lead into the other.
3. Possibly a wacky idea 😄 since we can pass a gameserver.yaml to the local game server for state -- could we make the "out of cluster mode" server create the `agones.dev/dev-address` for you? It's using local kubectl permissions anyway -- save you a `kubectl apply`? Just trying to streamline development even more.

Reply:
> I'm _almost_ wondering if this would be better as a part of https://agones.dev/site/docs/guides/local-game-server/ than here 🤔

> I'm thinking we may want this in https://agones.dev/site/docs/guides/local-game-server/ as well

Ummm, this is https://agones.dev/site/docs/guides/local-game-server/ (`site/content/en/docs/Guides/local-game-server.md`). Maybe I'm misunderstanding you? 😅

> (at the very least, they should probably link to each other though).

I added links to both pages, this is the one from https://agones.dev/site/docs/guides/local-game-server/ to https://agones.dev/site/docs/guides/client-sdks/local/. In `site/content/en/docs/Guides/Client SDKs/local.md` (line 180) I have the reverse link.

>  could we make the "out of cluster mode" server create the agones.dev/dev-address for you?

That seems like a large expansion of the behavior of the SDK Server. While it obviously isn't my call to say what's best for the codebase, it just seems like that change would be larger and more out of place.

Reply:
> Ummm, this is https://agones.dev/site/docs/guides/local-game-server/ (site/content/en/docs/Guides/local-game-server.md). Maybe I'm misunderstanding you? 

Sorry, I 100% wasn't clear. I was wondering it the whole new documentation set in `local.md` should be in this documentation - since my gut says that they are likely always going to be used together. WDYT? That also then solves any confusion around connection etc.

Also I was providing feedback generally to the docs, not just these specific pages -- understandably confusing.

> > could we make the "out of cluster mode" server create the agones.dev/dev-address for you?

> That seems like a large expansion of the behavior of the SDK Server. While it obviously isn't my call to say what's best for the codebase, it just seems like that change would be larger and more out of place.

What do you think of the idea though? Can always be a secondary issue filed for future work, just thought it might be handy.

Reply:
> I was wondering it the whole new documentation set in `local.md` should be in this documentation - since my gut says that they are likely always going to be used together.

I suppose it _could_ go in either place. Personally, I feel like this has _more_ to do with running the local SDK Server (as you need to change the commands used when running it). I definitely think "out of cluster" mode should be mentioned in both spots.

> What do you think of the idea though? Can always be a secondary issue filed for future work, just thought it might be handy.

If you're just asking for my opinion, I don't particularly see any use for it (for me). I suppose some might like omitting a `kubectl apply`, but that's not really what I'd be optimizing for. What I'd be looking for is "breakpoint locally but in the most prod-y way possible". As the expected order of lifetimes/creation on prod would be: `GameServer` resource, SDK Server binary, custom game server binary, I would want to keep it in that order. This means that the custom game binary, the allocation system, etc, can all be run in a consistent state (code abstracting away from this flow vs prod). Swapping around the order or operations, or changing what creates what, for dev testing (in such an integrated context) doesn't really make sense to me, as it isn't representative of how things will be working at the end.

Speaking of "other things that could be done for handy-ness":
As I alluded to in https://github.com/googleforgames/agones/pull/3252#discussion_r1259155061, I have other desires for potential future PRs. Namely that dev servers wouldn't automatically go from `Creating` to `Ready`. Instead, they would do a normal `Creating` to `Scheduled`, and would only progress to `Ready` after the SDK call was made (+ controller state progression) -- all of this would match https://agones.dev/site/docs/reference/gameserver/#gameserver-state-diagram. An `annotation` could be used to "auto-ready" (e.g. `agones.dev/auto-ready: true`) to keep the current functionality. Also an `annotation` for `agones.dev/recycle-gs: true` would be nice to progress from `Shutdown` to `Creating`/`Scheduled` (so when the binary exits, calling SDK Shutdown, the `GameServer` would just be reset instead of deleted). Again, all of this is a bit different/extra -- this PR is just trying to add in the finishing touches for a very useful flow that's mostly there already. Lol, I hesitated to really go into all this extra stuff as I don't want it to distract from this (seemingly more straightforward) PR. Hopefully this helps with the perspective that I'm looking at things from.

Reply:
These are cool ideas! I'd love you to write up some issues and designs for these and we can discuss them further.

Thinking about docs organisation some more, https://agones.dev/site/docs/guides/client-sdks/local/ is about a purely local experience (and the simplest to get started with).

This file is about integrating with your online system, likely tied to how you like to run things in production (also it's further down in the docs, so it's assumed by this point, you've managed to work out how the SDK works, done some integration etc). 

Jumping from "how do I integrate the SDK" to "now I'm working in a cluster" within the same page as it is now seems to me like a big leap concept with for new user onboarding to me.

So let's move the out-of-cluster documentation into a new section here, and add a link to this doc at the bottom (I'll add a suggestion to `local.md` on what I think will work there), since I think the annotations and the out-of-cluster mode is all interlinked (I don't see one being done without the other), without changing the learning path drastically.

Reply:
>  I'd love you to write up some issues and designs for these and we can discuss them further.

I went ahead and filed https://github.com/googleforgames/agones/issues/3284.

>  let's move the out-of-cluster documentation into a new section

I've now moved all the "out of cluster" documentation into `site/content/en/docs/Advanced/out-of-cluster-dev-server.md`. This actually allowed me to go much further into various details. I've updated `site/content/en/docs/Guides/Client SDKs/local.md` and `site/content/en/docs/Guides/local-game-server.md` with `Next Steps` that link to out of cluster configuration.